src/java/org/apache/cassandra/utils/bytecomparable/ByteComparable.md

```diff
@@ -0,0 +1,689 @@
+uses variable-length encoding, and the other doesn't. Longer numbers compare smaller when negative (because of the 
+inverted length bytes), and bigger when positive.
+
+Examples:
+
+|    value | bytes                   |encodes as|
+|---------:|-------------------------|---|
+|        0 | 00                      | 80
+|        1 | 01                      | 81
+|       -1 | FF                      | 7F
+|      255 | 00 FF                   | C0 FF
+|     -256 | FF 00                   | 3F 00
+|      256 | 01 00                   | C1 00
+|     2^16 | 01 00 00                | E1 00 00
+|    -2^32 | FF 00 00 00 00          | 07 00 00 00 00
+|   2^56-1 | 00 FF FF FF FF FF FF FF | FE FF FF FF FF FF FF FF
+|    -2^56 | FF 00 00 00 00 00 00 00 | 01 00 00 00 00 00 00 00
+|     2^56 | 01 00 00 00 00 00 00 00 | FF·00·01 00 00 00 00 00 00 00
+| -2^56-1  | FE FF FF FF FF FF FF FF | 00·FF·FE FF FF FF FF FF FF FF
+|   2^1024 | 01 00(128 times)        | FF·7A·01 00(128 times)
+|  -2^2048 | FF 00(256 times)        | 00·7F 06·00(256 times)
+
+(Middle dot · shows the transition point between length and digits.)
+
+## Variable-length floating-point decimals (decimal)
+
+Variable-length floats are more complicated, but we can treat them similarly to IEEE-754 floating point numbers, by
+normalizing them by splitting them into sign, mantissa and signed exponent such that the mantissa is a number below 1 
+with a non-zero leading digit. We can then compare sign, exponent and mantissa in sequence (where the comparison of
+exponent and mantissa are with reversed meaning if the sign is negative) and that gives us the decimal ordering.
+
+A bit of extra care must be exercised when encoding decimals. Since fractions like `0.1` cannot be perfectly encoded in
+binary, decimals (and mantissas) cannot be encoded in binary or base-256 correctly. A decimal base must be used; since 
+we deal with bytes, it makes most sense to make things a little more efficient by using base-100. Floating-point 
+encoding and the comparison idea from the previous paragraph work in any number base.
+
+`BigDecimal` presents a further challenge, as it encodes decimals using a mixture of bases: numbers have a binary-
+encoded integer part and a decimal power-of-ten scale. The bytes produced by a `BigDecimal` are thus not suitable for 
+direct conversion to byte comparable and we must first instantiate the bytes as a `BigDecimal`, and then apply the 
+class’s methods to operate on it as a number.
+
+We then use the following encoding:
+- If the number is 0, the encoding is a single `0x80` byte.
+- Convert the input to signed mantissa and signed exponent in base-100. If the value is negative, invert the sign of the
+  exponent.
+- Output a byte encoding:
+  - the sign of the number encoded as `0x80` if positive and `0x00` if negative,
+  - the exponent length (stripping leading 0s) in bytes as `0x40 + exponent_length * exponent_sign`.
```

Comment:
Reading the code makes it more clear that this first byte is `sign bit inverted * 0x80 + 0x40 + signed exponent length, where exponent is negated if value is negative`, but these two bullet points leave that a little ambiguous. Perhaps something like...

> Output a byte encoding the `signed exponent` + `0x40` + `signed exponent length`, where...
...the `signed exponent` is `0x80` if positive or `0x00` if negative, and...
...`signed exponent length` is the length of the exponent in bytes after stripping leading 0s